Define security around Window, WindowProxy, and Location properly

annevk opened the pull request:
This is a relative large change to the way Window, WindowProxy, and
Location objects work, at least from a standards’ perspective. This
should more closely match implementations and define all the relevant
details.

A rough summary of the changes:

* Window indexed getters have moved to WindowProxy.
* WindowProxy is now a JavaScript exotic object and it handles all the
security aspects for Window so Window can become an ordinary object.
* Cross-origin named properties for Window are now exposed (though on
WindowProxy). This addresses #544 and
https://www.w3.org/Bugs/Public/show_bug.cgi?id=21674 which that PR was
intended to replace.
* Location is now an exotic object with all its internal methods
overridden to handle the cross-origin security aspects.
* Location no longer uses Unforgeable on the interface, allowing that
to be removed from IDL. (It moved to all of its properties instead and
the remaining details are defined through prose.)

This should also address these bugs:

* https://www.w3.org/Bugs/Public/show_bug.cgi?id=20701
* https://www.w3.org/Bugs/Public/show_bug.cgi?id=27502
* https://www.w3.org/Bugs/Public/show_bug.cgi?id=22346
* https://www.w3.org/Bugs/Public/show_bug.cgi?id=27128

annevk commented:
Things I need to look into:

* Now `WindowProxy` handles indexed properties directly, we should expose them through [[OwnPropertyKeys]] too. Currently we only do that for the cross-origin case.
* I need to get https://github.com/tc39/ecma262/pull/356 landed and add some references.

Things that made me ponder:

* Although I used `<span>` around a bunch of Ordinary* abstract operation references, the build script is not complaining about missing `<dfn>`s. What is up with that?

In any event, this is definitely ready for review.

annevk commented:
If https://github.com/tc39/ecma262/pull/367 lands we need to figure out what to do with [[Enumerate]]. It's not entirely clear to me what we can do then.

domenic commented:
> Although I used \<span> around a bunch of Ordinary* abstract operation references, the build script is not complaining about missing \<dfn>s. What is up with that?

I've noticed this too. I think the build script thinks that empty spans are just meant to be using span to mark up a length of text, not auto-linking. That is why we should move to \<a> longer term.

Will try to review Monday.

sideshowbarker commented:
>> Although I used \<span> around a bunch of Ordinary* abstract operation references, the build script is not complaining about missing \<dfn>s. What is up with that?

> I've noticed this too. I think the build script thinks that empty spans are just meant to be using span to mark up a length of text, not auto-linking. That is why we should move to \<a> longer term.

To be clear, that’s an issue with wattsi, right? (Not something else in the `build.sh` script or the Perl scripts it calls.) If so, I wonder if https://github.com/whatwg/wattsi/pull/5 regressed this, or if it’s always been this way, or what.

Ms2ger commented on source:
```diff
@@ -77384,33 +77672,18 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+  <p>The <dfn>number of child browsing contexts</dfn> is the number of <span data-x="child browsing
+  context">child browsing contexts</span> that are <span data-x="browsing context nested
+  through">nested through</span> elements that are <span data-x="in a document">in the
+  <code>Document</code></span> that is the <span>active document</span> of the <code>Window</code>
```
A `Window` does not have an "active document"; a browsing context does.

Ms2ger commented on source:
```diff
@@ -77757,31 +78033,203 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
-  <span>browsing context</span> had a single <code>Window</code> object, while still keeping
-  separate <code>Window</code> objects for each <code>Document</code>.</p>
+  <p>Every <code>WindowProxy</code> object has a
+  [[<dfn data-x="concept-windowproxy-window">Window</dfn>]] internal slot representing the wrapped
```
`[[Window]]` is never set, only read.

domenic replied:
+1, setting [[Window]] on navigation is pretty important :). Seems like it's in "Initialising a new Document object", step 2. Also "create a new browsing context", step 1.

annevk replied:
Thank for you for pointing out the locations, fixup1 has the fix.

annevk commented:
WindowProxy needs to use https://heycam.github.io/webidl/#dfn-named-property-visibility (or a variant thereof) where it implements cross-origin named properties. In particular I think we don't want to have `[OverrideBuiltins]` behavior if I understood @bzbarsky's comments in https://bugzilla.mozilla.org/show_bug.cgi?id=1245554 correctly.

bzbarsky commented:
We should probably test what UAs actually do with cross-origin named properties carefully.  Defining how this should behave will be a bit tricky.  For example, consider this testcase:

    <!DOCTYPE html>
    <body>
      <iframe name="foo"></iframe>
      <script>
        Object.prototype.foo = 5;
        onload = function() { alert(window.foo); }
      </script>
    </body> 

In browsers this alerts "5" (at least in Chrome, Safari, and Firefox).  What should happen on cross-origin access to the "foo" property on this window?  Chrome and Firefox return the subframe, looks like.  Safari returns undefined (or 5, depending on what the accessing thing actually is; yes, it leaks info cross-origin in some cases).  Of course maybe Safari returns undefined for cross-origin named access in general?  We really need some exhaustive tests.  :(

domenic commented on source:
```diff
@@ -2976,16 +2977,23 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
      <li>The <dfn data-noexport="" data-x="js-prod-Module" data-x-href="https://tc39.github.io/ecma262/#prod-Module"><i>Module</i></dfn> production</li>
      <li>The <dfn data-noexport="" data-x="js-prod-Pattern" data-x-href="https://tc39.github.io/ecma262/#prod-Pattern"><i>Pattern</i></dfn> production</li>
      <li>The <dfn data-noexport="" data-x="js-prod-Script" data-x-href="https://tc39.github.io/ecma262/#prod-Script"><i>Script</i></dfn> production</li>
+     <li><dfn data-noexport="" data-x-href="https://tc39.github.io/ecma262/#sec-property-descriptor-specification-type">PropertyDescriptor</dfn></li>
```
I've been curating this section, perhaps a bit obsessively. This should be "The [PropertyDescriptor] specification type"

annevk replied:
This is meant to refer to the tag literal descriptions of Property Descriptor records.

domenic replied:
Yeah but ES conflates them

domenic replied:
Still would like to see this fixed. The fact that ES uses specification type names as tags when creating literals is just notation; the actual thing being referenced is the PropertyDescriptor specification type.

annevk replied:
Should it then say "Property Descriptor specification type"? But the references can use the literal PropertyDescriptor?

domenic replied:
Yep.

domenic commented on source:
```diff
@@ -2976,16 +2977,23 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
      <li>The <dfn data-noexport="" data-x="js-prod-Script" data-x-href="https://tc39.github.io/ecma262/#prod-Script"><i>Script</i></dfn> production</li>
+     <li><dfn data-noexport="" data-x-href="https://tc39.github.io/ecma262/#sec-property-descriptor-specification-type">PropertyDescriptor</dfn></li>
      <li>The <dfn data-noexport="" data-x-href="https://tc39.github.io/ecma262/#sec-source-text-module-records">Source Text Module Record</dfn> specification type and its <dfn data-noexport="" data-x="js-ModuleEvaluation" data-x-href="https://tc39.github.io/ecma262/#sec-moduleevaluation">ModuleEvaluation</dfn> and <dfn data-noexport="" data-x="js-ModuleDeclarationInstantiation" data-x-href="https ...
+     <li><dfn data-noexport="" data-x-href="https://tc39.github.io/ecma262/#current-realm">The current Realm Record</dfn></li>
```
This should move up to the "untyped" section at the top

annevk replied:
I'm not sure what this means, but I made a best guess.

domenic commented on source:
```diff
@@ -2976,16 +2977,23 @@ a.setAttribute('href', 'http://example.com/'); // change the content attribute d
+     <li>The <dfn data-noexport="" data-x-href="https://tc39.github.io/ecma262/#sec-isdatadescriptor">IsDataDescriptor</dfn> abstract operation</li>
      <li>The <dfn data-noexport="" data-x="js-NewObjectEnvironment" data-x-href="https://tc39.github.io/ecma262/#sec-newobjectenvironment">NewObjectEnvironment</dfn> abstract operation</li>
-     <li>The <dfn data-noexport="" data-x="js-ParseModule" data-x-href="https://tc39.github.io/ecma262/#sec-parsemodule">ParseModule</dfn> abstract operation
+     <li>The <dfn data-noexport="" data-x-href="https://tc39.github.io/ecma262/#sec-ordinarydefineownproperty">OrdinaryDefineOwnProperty</dfn> abstract operation</li>
```
Add in all the other OrdinaryXs. Consider sticking with the js- prefix, even though it's kind of a pain and I might take it back if I were starting over.

annevk replied:
Perhaps it's better to move away from it over time?

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <h4>Integration with IDL</h4>
+
+  <p>When <span>perform a security check</span> is invoked, with a <var>platformObject</var>,
```
This isn't linking anywhere

annevk replied:
This is fixed by fixup1.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+        <b>true</b>, then:</p>
+
+        <ol>
+         <li><p>If <var>type</var> is "method" and <var>e</var> has neither [[needsGet]] nor
```
Do we want to do the typical `"<code data-x="">method</code>"` style here? Or just leave it un-coded?

annevk replied:
I went with IDL's precedent, but since we're not copying ECMAScript's precedent I'll use `<code>`. Hopefully at some point we'll get all standards to align, since this is just unnecessarily confusing readers.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+      <ol>
+       <li>
+        <p>If <span>SameValue</span>(<var>e</var>.[[property]], <var>identifier</var>) is
+        <b>true</b>, then:</p>
```
HTML does not typically bold "true" (or "null"). In the few ES-spec-ese sections I've written (HostResolveImportedModule and HostPromiseRejectionTracker), I've stuck with not-bold.

annevk replied:
Okay, removed.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   </li>
+
+   <li><p>If <var>platformObject</var>'s global object's <span>effective script origin</span> is not
+   <span>same origin</span> with <span>the current Realm Record</span>'s global object's
```
s/global object/[[globalObject]]/

domenic replied:
I can't find where global objects get effective script origins, only Documents. Maybe you need to add "responsible document" to the chain??

annevk replied:
I guess I should start using "relevant settings object" here now that we changed to work for platform objects as well. It still doesn't feel like the best of setups though.

See also the discussion in https://www.w3.org/Bugs/Public/show_bug.cgi?id=28375.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  <!-- we should probably remove the realm argument from IDL -->
+
+
+  <h4>Shared internal slot: [[<span>crossOriginPropertyDescriptorMap</span>]]</h4>
```
I think the link in the heading looks bad and doesn't really add much since it's just linking a sentence down.

annevk replied:
HTML does this all the time though.

domenic replied:
Links close by, yes, but in headings?

annevk replied:
E.g., "The Window object".

domenic replied:
Yeah, for `<code>`s it is. But e.g. for "Browsing contexts" (and all of its subheadings, which also have terms corresponding to them) it is not.

annevk replied:
Ok, will remove the `<span>`.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <p><code>Window</code> and <code>Location</code> objects both have a
+  [[<dfn>crossOriginPropertyDescriptorMap</dfn>]] internal slot, whose value is initially an empty
+  list.
```
Capitalize List, and add an entry to the references section for the List specification type that you can cross-ref.

annevk replied:
This should say map.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  list.
+
+  <p>User agents should allow a value held in the map to be garbage collected along with its
+  corresponding key when nothing holds a reference to any part of the value. I.e., as long as
```
The hanging i.e. is a bit awkward. Maybe make it a parenthetical sentence: "(That is, as long as garbage collection is not observable.)"

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  corresponding key when nothing holds a reference to any part of the value. I.e., as long as
+  garbage collection is not observable.</p>
+
+  <p>E.g., with <code data-x="">const href = Object.getOwnPropertyDescriptor(crossOriginLocation,
```
Make this class="example" and rephrase to "For example, if author code did \<pre>...\</pre> the value... collected, as that would be observable"

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  "href").set</code> the value and its corresponding key in the map cannot be garbage collected as
+  that would be observable.</p>
+
+  <p>User agents may have an optimization whereby they remove key-value pairs from the map when
```
Another class="example" paragraph would be pretty great for this.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   object's <span>effective script origin</span>, and <b>false</b> otherwise.</p></li>
+  </ol>
+
+  <h5><dfn>CrossOriginGetPrototypeOf</dfn> ( <var>O</var> )</h5>
```
I think these need different names if they're planning to handle both same origin and cross-origin. WindowProxyOrLocationGetPrototypeOf? :-/

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  caller needs to throw a <code>SecurityError</code> exception.</p>
+
+  <ol>
+   <li><p>If <var>P</var> is <span>@@toStringTag</span>, <span>@@hasInstance</span>, or
```
The symbols aren't linking correctly

annevk replied:
This should be addressed by fixup1.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  data-x="dom-document-domain">document.domain</code> cannot revisit an earlier value.</p>
+
+
+  <h4>Shared abstract operations</h4>
```
This needs a note about how all of the abstract operations here apply only to ordinary objects.

domenic replied:
Meh, trying to figure out how to phrase such a note made me give up on the idea. Maybe a note later when defining WindowProxy. We'll see when I get there.

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>Return <b>undefined</b>.</p></li>
+  </ol>
+
+  <h5><dfn>CrossOriginPropertyDescriptor</dfn> ( <var>crossOriginProperty</var>,
```
I would nest all three of CrossOriginPropertyDescriptor, CrossOriginFunctionWrapper, and cross-origin wrapper functions under CrossOriginGetOwnPropertyHelper. (No need to nest cross-origin wrapper functions under CrossOriginFunctionWrapper. In fact maybe it's OK to not give it its own section.)

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   </li>
+
+   <li>
+    <p>Otherwise, then:</p>
```
Just "Otherwise,"

zcorpan replied:
Or `Otherwise:` to be consistent with other parts of the spec.

domenic commented on source:
```diff
@@ -76785,7 +77151,6 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
   readonly attribute <span>Element</span>? <span data-x="dom-frameElement">frameElement</span>;
   <span>WindowProxy</span>? <span data-x="dom-open">open</span>(optional DOMString url = "about:blank", optional DOMString target = "_blank", [TreatNullAs=EmptyString] optional DOMString features = "", optional boolean replace = false);
-  <span data-x="dom-window-item">getter</span> <span>WindowProxy</span> (unsigned long index);
   <span data-x="dom-window-namedItem">getter</span> object (DOMString name);
```
What do you think of adding a comment like "// indexed access to subframes is taken care of by WindowProxy" after this line?

domenic replied:
The named getter here means Window is exotic, which means you need to delegate to it per our earlier discussions instead of using OrdinaryX(). Did you mean to move the named getter elsewhere?

bzbarsky replied:
For global objects, the named getter adds an exotic NamedPropertiesObject on the proto chain instead of making the global itself exotic.

domenic replied:
That is some of the spookiest action at a distance I have seen in a while -_-. Maybe a comment? "// since this is a global, per IDL, this adds an exotic NamedPropertiesObject to the prototype chain"

annevk replied:
I have added a comment in fixup1 explaining both.

domenic replied:
Comment looks good. Maybe add a blank line between open() and the getter.

annevk replied:
We should do that separately, since currently each newline is indicative of a new section. So that would require some extra annotations here and there.

domenic commented on source:
```diff
@@ -77217,33 +77505,18 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+  <span>number of child browsing contexts</span>.</p>
 
-  <p>These properties are the <dfn>dynamic nested browsing context properties</dfn>.</p>
+  <p class="note">The indexed getter is defined through the
```
Not quite the right phrasing; there is no indexed getter, since that is an IDL concept. Maybe "indexed access to child browsing contexts is defined through..."

domenic commented on source:
```diff
@@ -77590,31 +77866,203 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
 
   <h4>The <code>WindowProxy</code> object</h4>
```
IMO this probably deserves to get bumped up a level (h3) now that it's not just a hand-wavy thing placed in front of Window.

domenic commented on source:
```diff
@@ -77590,31 +77866,203 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+   <b>this</b>.</p></li>
+
+   <li>
+    <p>If <var>P</var> is an <span>array index property name</span>, then:</p>
```
This and the subsequent step could probably use https://tc39.github.io/ecma262/#sec-canonicalnumericindexstring instead. I don't have a strong preference.

annevk replied:
That seems very different from https://heycam.github.io/webidl/#dfn-array-index-property-name though.

domenic replied:
Yeah, they do it a bit differently I guess; there's a later IsInteger check. Probably not worth it.

domenic commented on source:
```diff
@@ -77590,31 +77866,203 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+    <p>If <var>P</var> is an <span>array index property name</span>, then:</p>
+
+    <ol>
+     <li><p>Let <var>index</var> be ToUint32(<var>P</var>).</p></li>
```
ToUin32 needs to link

domenic commented on source:
```diff
@@ -77590,31 +77866,203 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+      <var>maxProperties</var> &minus; 1, then:
+
+      <ol>
+       <li><p>Set <var>value</var> to the <code>WindowProxy</code> object of the <var>index</var>th
```
Seriously, browsers!?!? This is the worst.

domenic commented on source:
```diff
@@ -77590,31 +77866,203 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
-  <div class="example">
+  <p>The <code>WindowProxy</code> object internal methods are described in the subsections
+  below.</p>
 
```
Here is where my NOTE goes.

> NOTE: although WindowProxy is named as a "proxy", it does not do polymorphic dispatch on its target's internal methods like a real proxy would, due to a desire to reuse machinery between Location and WindowProxy objects. As long as the Window object does not exhibit any exotic behaviors (such as indexed or named getters or setters) this is unobservable.

domenic commented on source:
```diff
@@ -77590,31 +77866,203 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+   [[<span data-x="concept-windowproxy-window">Window</span>]] internal slot of
+   <b>this</b>.</p></li>
+
+   <li><p>Return ? <span>CrossOriginGet</span>(this, <var>W</var>, <var>P</var>,
```
I don't understand why the second parameter is named proxyO if you're passing the Window, not the WindowProxy, as proxyO.

Again I think this would be much clearer if the "is same origin" check was inlined into the call sites instead of having to pass in these two different objects. Then you could keep the CrossOriginX names too.

domenic commented on source:
```diff
@@ -77590,31 +77866,203 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+   <li><p>Return <span>CrossOriginDelete</span>(<var>W</var>, <var>P</var>).</p></li>
+  </ol>
+
+  <h5 id="windowproxy-enumerate">[[Enumerate]] ( )</h5>
```
Enumerate is either dead or almost dead. Kill it here and elsewhere.

domenic commented on source:
```diff
@@ -79159,6 +79607,59 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   object</span>.</p></li>
+
+   <li><p>Perform ! <var>location</var>.[[DefineOwnProperty]]("toString", {
+   [[Value]]: %ObjProto_toString%,
```
Cross-ref %ObjProto_toString%

domenic commented on source:
```diff
@@ -79159,6 +79607,59 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   [[Enumerable]]: <b>false</b>,
+   [[Configurable]]: <b>false</b> }).</p></li>
+
+   <li><p>Perform ! <var>location</var>.[[DefineOwnProperty]](@@toPrimitive, { [
```
Cross link `@@toPrimitive`.

domenic commented on source:
```diff
@@ -79159,6 +79607,59 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
 
+  <div w-nodev>
+
+  <p>To create a <code>Location</code> object, run these steps:</p>
```
Might want something about how Location is a mish-mash of IDL with overwritten creation behavior and internal methods?

domenic commented on source:
```diff
@@ -79662,10 +80263,18 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
 
   </ol>
 
+  <p class="note">The <code data-x="dom-location-replace">replace(<var>url</var>)</code> method
```
Do we usually refer to methods including their parameter names? Or would this just be "The location method"?

annevk replied:
I made it "The replace() method" locally.

domenic commented on source:
```diff
@@ -79739,72 +80350,138 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   <li><p><i>End</i>: Return <var>output</var>.</p></li>
+  </ol>
+
+  <h5>Security infrastructure specific to the <code>Location</code> object</h5>
```
I think we don't need a new subsection here. Just an intro paragraph explaining how Location is a special pumpkin that overrides internal methods for security reasons.

domenic commented on source:
```diff
@@ -79739,72 +80350,138 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
-  changes to that <code>Location</code> object's properties, getters, setters, etc, were not
-  present, and as if all the properties of that <code>Location</code> object had their
-  [[Enumerable]] attribute set to false.</p>
+   <li><p>Let <var>property</var> be <span>CrossOriginGetOwnProperty</span>(<b>this</b>,
```
Missing "Helper" causes failed link

annevk replied:
Addressed.

domenic commented on source:
```diff
@@ -79739,72 +80350,138 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+  <h6 id="location-set-v-receiver">[[Set]] ( <var>P</var>, <var>V</var>, <var>Receiver</var> )</h6>
+
+  <ol>
+   <li>Return <span>CrossOriginSet</span>(<b>this</b>, <b>this</b>, <var>P</var>, <var>V</var>,
```
Seeing these repeated "this, this" I feel even more strongly the security check should be pulled out. The fact that it's the same object for Location and the Window for WindowProxy is an important detail that gets lost when you only see "O, proxyO" in the abstract.

domenic commented on source:
```diff
@@ -79739,72 +80350,138 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+  <h6 id="location-delete">[[Delete]] ( <var>P</var> )</h6>
+
+  <ol>
+   <li>Return <span>CrossOriginDelete</span>(this, <var>P</var>).</li>
```
Missing bold for `this` here.

Also, if we're going with ES-style bold `this`, it'd be good to fix line 75927 where I do `<code data-x="">this</code>`. I think that's the only other place in the spec though, based on searching for `>this<`.

annevk replied:
I found another one around 85k. Also changed. (Not yet pushed.)

domenic commented on source:
```diff
@@ -76751,7 +76759,365 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
 
 
-<!--TOPIC:DOM APIs-->
+  <h3 id="cross-origin-objects">Security infrastructure for <code>Window</code>,
```
Maybe Window doesn't belong on this list?

annevk replied:
It does, due to the IDL security check happening on Window, not WindowProxy.

annevk commented:
Outstanding feedback, ignoring tests for the moment which I'd like to look into after we have a standard to write them against:

* Now WindowProxy handles indexed properties directly, we should expose them through [[OwnPropertyKeys]] too. Currently we only do that for the cross-origin case.
* [[Enumerate]] needs to be removed, but to do that GOPD (in the cross-origin case) always needs to return false for [[Enumerable]] and GetPrototypeOf needs to return null. The latter is already the case.
* I would nest all three of CrossOriginPropertyDescriptor, CrossOriginFunctionWrapper, and cross-origin wrapper functions under CrossOriginGetOwnPropertyHelper. (No need to nest cross-origin wrapper functions under CrossOriginFunctionWrapper. In fact maybe it's OK to not give it its own section.)
* Bump WindowProxy up to `<h3>`
* Remove Location subsection in favor of an intro paragraph explaining how Location is a special pumpkin that overrides internal methods for security reasons. Perhaps `<hr>` could be added though.
* Domenic wants the security check pulled out primarily because of "O, proxyO". Investigate if that's feasible.

annevk commented:
Another comment was renaming CrossOrigin* to WindowProxyOrLocation*, which we should do if we don't pull out the security check.

domenic commented:
> Domenic wants the security check pulled out primarily because of "O, proxyO". Investigate if that's feasible.

To expand on this a bit, I think it's much easier to understand if "SomethingSomethingGet" takes the same arguments as ES's "Get". It's strange now how some match (Delete, HasProperty) but others don't (Get, Set).

And, I think it's clearer to show how the security checks are on very different objects if they're present in each definition, instead of using this, this for Location and this, W for WindowProxy, then having a generic O vs. proxyO in the operation.

annevk commented:
@bholley, @bzbarsky, if you are curious I have uploaded a single-page version of the standard that has this PR merged here: https://html5.org/temp/cross-origin-objects-review.html#cross-origin-objects.

Other interesting bits are https://html5.org/temp/cross-origin-objects-review.html#windowproxy and https://html5.org/temp/cross-origin-objects-review.html#the-location-interface (internal methods at the end of this section).

annevk commented:
I believe I have addressed all feedback thus far, but I'm in no rush to merge this so I'll wait at least until Tuesday next week, but let me know if you want more time.

annevk commented:
@ajklein ^^

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  to "<code data-x="">example.com</code>" on www.example.com, means user agents can remove all
+  key-value pairs from the map where part of the key is www.example.com as that can never be part of
+  the <span>effective script origin</span> again and therefore the corresponding value could never
+  be retrieved from the map.</p>
```
Move a comma around: "on example.com means user agents ... key is www.example.com, as that can..."

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>Assert: <var>O</var> is a <code>Location</code> or <code>Window</code> object.</p></li>
+
+   <li><p>If <var>O</var> is a <code>Location</code> object, then return «
+   { [[property]]: "href", [[needsGet]]: false, [[needsSet]]: true },
```
These strings need `<code>`-ifying

annevk replied:
Providing an xref would be weird though, right?

domenic replied:
Yeah, I agree it would be weird.

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+    <ol>
+     <li><p>Let <var>value</var> be <var>originalDesc</var>.[[Value]].</p></li>
+
+     <li><p>If IsCallable(<var>value</var>) is true, set <var>value</var> to
```
IsCallable needs to link

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+   <li><p>Let <var>wrappedFunction</var> be the [[Wrapped]] internal slot of <var>F</var>.</p></li>
+
+   <li><p>Return ? Call(<var>wrappedFunction</var>, this, <var>argumentsList</var>).</p></li>
```
Call needs to link

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>If <span>IsDataDescriptor</span>(<var>desc</var>) is true, then return
+   <var>desc</var>.[[Value]].</p></li>
+
+   <li><p>Otherwise, <span>IsAccessorDescriptor</span>(<var>desc</var>) must be true so, let
```
This seems to be a typo in the ES source; should be "true, so let". Will file an ES bug.

annevk replied:
Seems they decided to bikeshed: https://github.com/tc39/ecma262/issues/389...

annevk replied:
This is not yet addressed, pending tc39.

domenic replied:
I'd do "true, so let" pending their decision; what's there is just wrong.

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+   <li><p>Otherwise, <span>IsAccessorDescriptor</span>(<var>desc</var>) must be true so, let
+   <var>getter</var> be <var>desc</var>.[[Get]].</p></li>
+
+   <li><p>If <var>getter</var> is not undefined, return ? Call(<var>getter</var>,
```
Link Call

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+     <li><p>If <var>setter</var> is undefined, return false.</p></li>
+
+     <li><p>Perform ? Call(<var>setter</var>, <var>Receiver</var>, «<var>V</var>»).</p></li>
```
Link Call

domenic commented on source:
```diff
@@ -77505,33 +77766,239 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+  <p>A <dfn><code>WindowProxy</code></dfn> is an exotic object that wraps a <code>Window</code>
+  object, indirecting most operations through to the wrapped object. Each
+  <span>browsing context</span> has an associated <code>WindowProxy</code> object. When the
+  <span>browsing context</span> is navigated, the <code>Window</code> object wrapped by the
```
Maybe navigated should be a link.

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+  </ol>
+
+
+  <h4>Shared internal slot: [[crossOriginPropertyDescriptorMap]]</h4>
```
I now realize that this should probably be capitalized ([[CrossOriginPropertyDescriptorMap]]). In streams I for some reason went with initial lowercase, but most of ES (not all) uses initial-uppercase, and so do we for [[Window]].

annevk replied:
Does this also apply to fields, such as [[needsGet]]?

annevk replied:
This is not yet addressed, wanted to know the scope first.

domenic replied:
Good point. Yes, those should probably be uppercase too.

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <h4>Shared abstract operations</h4>
+
+  <h5><dfn>CrossOriginProperties</dfn>( <var>O</var> )</h5>
```
Need space between CrossOriginProperties and (

domenic commented on source:
```diff
@@ -76668,7 +76695,326 @@ dictionary <dfn>DragEventInit</dfn> : <span>MouseEventInit</span> {
+
+  <p><code>Window</code> and <code>Location</code> objects both have a
+  [[<dfn>crossOriginPropertyDescriptorMap</dfn>]] internal slot, whose value is initially an empty
+  map.
```
Maybe it would be good to add a sentence explaining what it is used for in practice, so people can get an intuition for it. This might not be right, but I think it's something like:

This contains entries mapping from (_origin1_, _origin2_, _property key_)-tuples to property descriptors, as a memoization of what is visible to author scripts when the _origin1_ inspects a `Window` or `Location` object from _origin2_. It is filled lazily by [CrossOriginGetOwnPropertyHelper], which consults it on future lookups.

annevk replied:
I added a note in fixup2.

domenic commented on source:
```diff
@@ -77505,33 +77766,239 @@ callback <dfn>FrameRequestCallback</dfn> = void (<span>DOMHighResTimeStamp</span
+    <p>Repeat while <var>index</var> &lt; <var>maxProperties</var>,
+
+    <ol>
+     <li><p>Add ! ToString(<var>index</var>) as the last element of <var>keys</var>.</p></li>
```
Link ToString

domenic commented on source:
```diff
@@ -79016,6 +79483,64 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   <li><p>Let <var>location</var> be a new <code>Location</code> <span>platform
+   object</span>.</p></li>
+
+   <li><p>Perform ! <var>location</var>.[[DefineOwnProperty]]("toString", {
```
Code-ify strings throughout this algorithm

domenic commented on source:
```diff
@@ -79016,6 +79483,64 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   [[Configurable]]: false }).</p></li>
+
+   <li><p>Set the value of the [[<span>defaultProperties</span>]] internal slot of
+   <var>location</var> to <var>location</var>.[[OwnPropertyKeys]]().</p></li>
```
Add a ! before _location_.[[OwnPropertyKeys]]().

domenic replied:
Actually not really needed; ES is a bit lax about conflating completion values with the normal completions they represent, on purpose. Otherwise you'd have to replace all the IsXyz()s with ! IsXyz()s.

domenic commented on source:
```diff
@@ -79211,9 +79736,18 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   <li><p>If this <code>Location</code> object's <span>relevant <code>Document</code></span>'s
+   <span>effective script origin</span> is not <span>same origin</span> with <span>entry settings
+   object</span>'s <span>effective script origin</span>, throw a <code>SecurityError</code>
+   exception.</p></li>
```
I don't suppose IsPlatformObjectSameOrigin would work here, would it?

annevk replied:
Entry settings object is not the current Realm Record's [[globalObject]]'s relevant settings object, so I don't think so.

domenic commented on source:
```diff
@@ -79579,72 +80214,151 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+  <code>Location</code> objects must implement are defined below.</p>
+
+  <p>Every <code>Location</code> object has a [[<dfn>defaultProperties</dfn>]] internal slot
+  representing the own properties at time of creation.</p>
```
"representing *its* own properties at *the* time of *its* creation"

annevk replied:
I guess we should make this [[DefaultProperties]] then per the earlier comment?

domenic commented on source:
```diff
@@ -79579,72 +80214,151 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+
+  <ol>
+   <li><p>If <span>IsPlatformObjectSameOrigin</span>(<b>this</b>) is true, then return ?
+   <span>OrdinaryHasProperty</span>(<b>O</b>, <var>P</var>).</p></li>
```
s/O/this

ajklein commented:
@verwaest for more v8 WindowProxy expertise

domenic commented:
Finished another pass. Pretty much all editorial stuff. This is looking really good. I'm impressed and excited.

annevk commented:
https://html5.org/temp/cross-origin-objects-review.html now hosts the latest version, including fixup2.

annevk commented:
Supporting indexed properties directly on `WindowProxy` objects is not as easy as I thought it would be. In particular, [[HasProperty]], [[Get]], [[Set]], and [[Delete]] all need another look and likely fixes for the same-origin case. (It also bothers me somewhat that IDL and ECMAScript seemingly have slightly different algorithms for dealing with indexed properties. I guess I'll try to side with IDL for now and file an issue of sorts.)

annevk commented:
Hopefully fixed now. URL above still contains the latest version.

annevk commented:
I filed https://github.com/tc39/ecma262/issues/393 as an editorial issue against ECMAScript that might influence what casing we use for internal records and fields. I also filed https://github.com/tc39/ecma262/issues/394 about the guarantees around Ordinary* abstract operations.

Review would still be appreciated, but I also think landing this would be okay at this point as a significant improvement over the status quo.

domenic commented:
- New question: why is Location's valueOf defined as a new function unique to each Location instance, but its toString gets to reuse %ObjProto_toString%?

- Consider replacing "If IsLocationDefaultProperty(**this**, _P_)" with just "If the [[DefaultProperties]] internal slot of **this** contains _P_".

Otherwise LGTM! We should merge this soon.

annevk commented:
`valueOf` and `toString` for `Location` are based on [Unforgeable] interfaces (which was designed for `Location`, but `Location` will no longer use and can therefore be removed from IDL (I filed a bug)): https://heycam.github.io/webidl/#es-platform-objects. Perhaps you are correct though and we should create %objProto_valueOf% in ECMAScript and use that here. @bzbarsky, what do you think?

annevk commented:
@heycam you might be able to help out with the question about valueOf on [Unforgeable] interfaces above too.

bzbarsky commented:
> Perhaps you are correct though and we should create %objProto_valueOf% in ECMAScript and use that here.

We could do that, I guess.  Note, however, that %objProto_valueOf% and the function defined in this diff both have different behavior from the valueOf function Web IDL defines for Location right now and that Gecko implements.  That function is defined to return `this`, while the new setup is returning `ToObject(this)`.  What's the reason for the behavior change?

bzbarsky commented:
Oh, and the valueOf thing is pretty inconsistent across browsers right now.  Safari doesn't have it as an own prop on Location.  Chrome has it as an own prop, but it enforces "this" being a Location, apparently, instead of just returning `this`.

annevk commented:
@bzbarsky I had assumed IDL was sloppy and intended to match `Object.prototype.valueOf`. Happy to change this back to what IDL used to say (which would also rule out uplifting this to ECMAScript).

domenic commented:
Since it's so inconsistent anyway, I'd prefer to go with the most sane thing (just making it a pointer to %ObjProto_valueOf%) instead of picking one vendor's idiosyncratic choice and canonizing that. Even if that is what IDL currently defines.

bzbarsky commented:
I am probably OK with that if there is some sort of guarantee that `ToObject()` on an object (and in general, the behavior of `Object.prototype.valueOf`) will never become hookable in ES.  That probably requires an explicit note in the ES spec indicating that it must never become so.  Because if it ever does become hookable, that introduces a security problem, and I have no expectation that an ES editor would know enough about `Location` to realize they're introducing one, unless the text of the ES spec explicitly points this out....

annevk commented:
The PR above includes such a comment in the ToObject section.

bzbarsky commented:
Should add one in the actual valueOf section too, probably...

annevk commented:
I'm having a hard time persuading folks that notifying downstream of these hooks is important. Even if I don't persuade them though, I think we should still go with %ObjProto_valueOf%. Any kind of hook ECMAScript adds anywhere poses a problem for these objects.

domenic commented:
OK, now that ES has the appropriate hook, this LGTM!!!!! I'll let you do the honors, @annevk. wooohoooo!!!

annevk commented:
@domenic I had a few more issues opened against ECMAScript that are now resolved (two PRs still pending). The biggest change is the removal of [[HasProperty]] overrides due to an upstream change. Since OrdinaryHasProperty will invoke [[GetOwnProperty]] now instead of OrdinaryGetOwnProperty.

annevk commented:
In another PR @domenic noted:

> I think for these kind of informal maps, I much prefer the phrasing I used for the module map:
>
> - If _module map_ contains an entry with key _url_, ...
> - If _module map_ contains an entry with key _url_ whose value is `"fetching"`, ...
> - Create an entry in _module map_ with key _url_ and value `"fetching"`

Pretty sure we're not using that terminology for the internal map we are introducing here.

domenic commented on source:
```diff
@@ -79525,72 +80143,127 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   <li><p>Return false.</p></li>
+  </ol>
+
+  <h5 id="location-get-receiver">[[Get]] ( <var>P</var>, <var>Receiver</var> )</h5>
```
Can this be "location-get"?

domenic commented on source:
```diff
@@ -79525,72 +80143,127 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+   <var>Receiver</var>).</p></li>
+  </ol>
+
+  <h5 id="location-set-v-receiver">[[Set]] ( <var>P</var>, <var>V</var>, <var>Receiver</var> )</h5>
```
"location-set"?

annevk commented:
Is the "do not merge yet" label there because of https://github.com/tc39/ecma262/pull/416?

domenic commented:
Yep

domenic commented:
Dependencies merged; land at will @annevk!

sideshowbarker commented:
Rock n rollーVery cool to see this landing.

domenic commented:
This might deserve a blog post :D

Ms2ger commented:
https://blog.whatwg.org/windowproxy-window-and-location

handonam commented on source:
```diff
@@ -78975,6 +79438,51 @@ State: &lt;OUTPUT NAME=I>1&lt;/OUTPUT> &lt;INPUT VALUE="Increment" TYPE=BUTTON O
+  <p class="warning">The <code>Location</code> exotic object is defined through a mishmash of IDL,
+  invocation of JavaScript internal methods post-creation, and overridden JavaScript internal
+  methods. Coupled with its scary security policy, please take extra care while implementing
+  this excrescence.</p>
```
Wanted to say this is such an incredible word choice that had me search for over an hour on how this was added. Now i can relieve myself.
